timelock-agent/src/main/java/com/palantir/atlasdb/timelock/paxos/ClientPaxosResourceFactory.java

```diff
@@ -117,19 +124,154 @@ private static NetworkClientFactories factories(
+            PaxosComponents components = useCaseContext.components();
+            BatchPaxosAcceptorResource acceptorResource = new BatchPaxosAcceptorResource(components.batchAcceptor());
+            BatchPaxosLearnerResource learnerResource = new BatchPaxosLearnerResource(components.batchLearner());
+            return ImmutableBatchPaxosResources.of(acceptorResource, learnerResource);
+        }
+    }
+
+    @Value.Immutable
+    public interface TimelockPaxosInstallationContext {
+
+        @Value.Parameter
+        TimeLockInstallConfiguration install();
+
+        @Value.Derived
+        default UUID nodeUuid() {
+            return UUID.randomUUID();
+        }
 
         @Value.Derived
-        default List<Closeable> closeables() {
-            return networkClientFactories().closeables();
+        default int quorumSize() {
+            return PaxosRemotingUtils.getQuorumSize(clusterAddress());
+        }
+
+        @Value.Derived
+        default Set<String> clusterAddress() {
+            return PaxosRemotingUtils.getClusterAddresses(install());
+        }
+
+        @Value.Derived
+        default Set<String> remoteUris() {
+            return PaxosRemotingUtils.getRemoteServerPaths(install());
+        }
+
+        @Value.Derived
+        default Path dataDirectory() {
+            return install().paxos().dataDirectory().toPath();
+        }
+
+        @Value.Derived
+        default Optional<TrustContext> trustContext() {
+            return PaxosRemotingUtils
+                    .getSslConfigurationOptional(install())
+                    .map(SslSocketFactories::createTrustContext);
+        }
+
+    }
+
+    @Value.Immutable
+    public abstract static class PaxosRemoteClients {
+
+        @Value.Parameter
+        public abstract TimelockPaxosInstallationContext context();
+
+        @Value.Parameter
+        public abstract TaggedMetricRegistry metrics();
+
+        @Value.Derived
+        public List<TimelockPaxosAcceptorRpcClient> nonBatchAcceptor() {
+            return createInstrumentedRemoteProxies(TimelockPaxosAcceptorRpcClient.class, "paxos-acceptor-rpc-client");
+        }
+
+        @Value.Derived
+        public List<TimelockPaxosLearnerRpcClient> nonBatchLearner() {
+            return createInstrumentedRemoteProxies(TimelockPaxosLearnerRpcClient.class, "paxos-learner-rpc-client");
+        }
+
+        @Value.Derived
+        public List<BatchPaxosAcceptorRpcClient> batchAcceptor() {
+            return createInstrumentedRemoteProxies(
+                    BatchPaxosAcceptorRpcClient.class,
+                    "batch-paxos-acceptor-rpc-client");
+        }
+
+        @Value.Derived
+        public List<BatchPaxosLearnerRpcClient> batchLearner() {
+            return createInstrumentedRemoteProxies(
+                    BatchPaxosLearnerRpcClient.class,
+                    "batch-paxos-learner-rpc-client");
+        }
```

Comment:
Following your concern on avoiding unnecessary creation: we _technically_ could have these 4 be `@Value.Lazy` and only initialize the proxies on first usage, the idea being that if `useBatchPaxos` is always true then you don't care about the unbatched ones. 

However, we already create all of them (line 113 onward). It's possible to change DynamicDecoratingProxy to lazily initialize stuff, but it seems like a large change for what I suspect is questionable gain.

Reply:
yeah, it was more of a "am I instantiating the world each time to create something per client?" I feel like I've pinned down most of them.